### PR TITLE
Fix scheduling for splits with locality requirement in Tardigrade

### PR DESCRIPTION
**Pull request by arhimondr**
## Description

Fixes several problems related to scheduling of non remotely accessible splits

> Is this change a fix, improvement, new feature, refactoring, or other?

Fix

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

Core engine (Tardigrade)

> How would you describe this change to a non-technical end user or system administrator?

Fixes scheduling for non remotely accessible splits in certain corner cases. Prior this fix in some queries scanning over non remotely accessible splits might've been failing. 

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

() No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Tardigrade
* Fix scheduling for non remotely accessible splits 
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/FullNodeCapableNodeAllocatorService.java [hunk 1]
```diff
@@ -512,7 +511,15 @@ private Candidates selectCandidates(NodeRequirements requirements)
         return new Candidates(
                 allNodes.size(),
                 allNodes.stream()
-                        .filter(node -> requirements.getAddresses().isEmpty() || requirements.getAddresses().contains(node.getHostAndPort()))
+                        .filter(node -> {
+                            if (requirements.getAddresses().contains(node.getHostAndPort())) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/FullNodeCapableNodeAllocatorService.java:515]: nit Add comment:
```
// allow using a coordinator if explicitly requested by address
```

File: plugin/trino-tpcds/src/main/java/io/trino/plugin/tpcds/TpcdsNodePartitioningProvider.java [hunk 2]
```diff
@@ -53,10 +55,17 @@ public TpcdsNodePartitioningProvider(NodeManager nodeManager, int splitsPerNode)
     public ConnectorBucketNodeMap getBucketNodeMap(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorPartitioningHandle partitioningHandle)
     {
         Set<Node> nodes = nodeManager.getRequiredWorkerNodes();
-        checkState(!nodes.isEmpty(), "No TPCDS nodes available");
-
-        // Split the data using split and skew by the number of nodes available.
-        return createBucketNodeMap(toIntExact((long) nodes.size() * splitsPerNode));
+        checkState(!nodes.isEmpty(), "No TPCH nodes available");
```
Comment (losipiuk) [inline plugin/trino-tpcds/src/main/java/io/trino/plugin/tpcds/TpcdsNodePartitioningProvider.java:58]: TPCDS

File: plugin/trino-tpcds/src/main/java/io/trino/plugin/tpcds/TpcdsNodePartitioningProvider.java [hunk 3]
```diff
@@ -53,10 +55,17 @@ public TpcdsNodePartitioningProvider(NodeManager nodeManager, int splitsPerNode)
...
+        List<Node> sortedNodes = nodes.stream()
```
Comment (losipiuk) [inline plugin/trino-tpcds/src/main/java/io/trino/plugin/tpcds/TpcdsNodePartitioningProvider.java:59]: add a comment why sorting is needed (for sake of future readers). And link to TpcdsSplitManager.
Add cross comment to.

And also in Tpch*

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java [hunk 4]
```diff
@@ -1900,16 +1902,28 @@ private static BucketToPartition createBucketToPartitionMap(
...
-                for (int bucket = 0; bucket < bucketCount; bucket++) {
-                    bucketToPartition[bucket] = nextPartitionId++ % partitionCount;
+                if (bucketNodeMap.isDynamic()) {
+                    int nextPartitionId = 0;
+                    for (int bucket = 0; bucket < bucketCount; bucket++) {
+                        bucketToPartition[bucket] = nextPartitionId++ % hashPartitionCount;
+                    }
+                }
+                else {
+                    BiMap<InternalNode, Integer> nodeToPartition = HashBiMap.create();
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:1915]: Why not just Map?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java [hunk 5]
```diff
@@ -1900,16 +1902,28 @@ private static BucketToPartition createBucketToPartitionMap(
...
+                    int nextPartitionId = 0;
+                    for (int bucket = 0; bucket < bucketCount; bucket++) {
+                        InternalNode node = bucketNodeMap.getAssignedNode(bucket)
+                                .orElseThrow(() -> new IllegalStateException("Nodes are expected to be assigned for non dynamic BucketNodeMap"));
+                        Integer partitionId = nodeToPartition.get(node);
+                        if (partitionId == null) {
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:1921]: would that make sense to have more than one partition on single node?

File: core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java [hunk 6]
```diff
@@ -32,7 +32,6 @@
 import io.trino.execution.TaskId;
 import io.trino.execution.TaskInfo;
 import io.trino.execution.TaskStatus;
-import io.trino.execution.scheduler.NodeSchedulerConfig;
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java:35]: nit:  change cm "Update coordinator memory pools" -> "Always update coordinator memory pools"

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java (hunk 4 above)
Comment (linzebing) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:1915]: I think it' beneficial to add a comment explaining the logic, e.g. "make sure all buckets mapped to the same node map to the same partition, such that locality requirements are respected in scheduling".

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java (hunk 5 above)
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:1921]: Hmm, interesting question. For example to make tasks smaller for more granular retries?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java [hunk 7]
```diff
@@ -439,8 +443,39 @@ public HashDistributionTaskSource(
                         int partition = getPartitionForBucket(bucket);
 
                         if (!bucketNodeMap.isDynamic()) {
-                            HostAddress existingValue = partitionToNodeMap.put(partition, bucketNodeMap.getAssignedNode(split).get().getHostAndPort());
-                            checkState(existingValue == null, "host already assigned for partition %s: %s", partition, existingValue);
+                            HostAddress requiredAddress = bucketNodeMap.getAssignedNode(split).get().getHostAndPort();
+                            Set<HostAddress> existingRequirement = partitionToNodeMap.get(partition);
```
[the remark this answers is not in the document]
Comment (arhimondr) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java:447]: A split has a list of hosts in it's requirement. The set is needed to support split specific requirements.

**Review by linzebing — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java (hunk 5 above)
Comment (losipiuk) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:1920]: Yeah for example. For partitions, we should probably opt for those to be similarly sized. Building partitions on top of the enforced bucket<->node mapping does not necessarily imply that. Not something that we need to address here. Just a random thought.

**Pull request merged**